[Dispatch Creation] Don't place bit-truncate in consumer dispatch

IanWood1 opened the pull request:
During `FormDispatchRegions` don't fuse truncate ops into the same dispatch as their consumers. Instead, they should be fused with their producers.

IanWood1 committed:
[Dispatch Creation] Don't place bit-truncate in consumer dispatch

Signed-off-by: Ian Wood <ianwood@u.northwestern.edu>

MaheshRavishankar reviewed (changes requested):
I think you want this in conjunction with your other PR that deals with elementwise op fusion right and have the same control lever?

IanWood1 committed:
Link to clEnableEarlyTruncFusion

Signed-off-by: Ian Wood <ianwood@u.northwestern.edu>

MaheshRavishankar commented on compiler/src/iree/compiler/DispatchCreation/Passes.td:
```diff
@@ -246,7 +246,9 @@ def FormDispatchRegionsPass :
     Option<"fusePadWithProducers", "fuse-pad-with-producers", "bool",
-           /*default=*/"false", "Enable fusion of pad with producers">
+           /*default=*/"false", "Enable fusion of pad with producers">,
+    Option<"fuseTruncWithConsumers", "fuse-trunc-with-producers", "bool",
```
I think you meant `fuseTruncWithProducers` here?

IanWood1 replied:
I think the name does what it says. It controls/prevents `trunc -> consumer` fusion in the hopes that the trunc will be fused with its producer. But it doesn't have direct control over if it will actually fuse with a producer.

MaheshRavishankar replied:
The `Option` has c++ variable name of "fuseTruncWithConsumer" and the flag as `fuse-trunc-with-producers`. That seems like a typo

MaheshRavishankar reviewed (approved):
Seems like there is a typo, but looks good otherwise.

IanWood1 committed:
Fix typo

Signed-off-by: Ian Wood <ianwood@u.northwestern.edu>

IanWood1 committed:
Run producer fusion twice

The second time we allow fusing with truncate operations. This ensures that
truncate ops with no producers still get fused instead of being in their
own dispatch.

Signed-off-by: Ian Wood <ianwood@u.northwestern.edu>

MaheshRavishankar reviewed (approved):
Saw the update on the PR already. Looks fine to me. Thanks for going the extra mile.
